SCons/Environment.py

```diff
@@ -2537,45 +2567,75 @@ def FindInstalledFiles(self):
+    base environment *subject*, except that same-named construction
+    variables from *overrides* are returned on read access; assignment
+    to a construction variable creates an override entry - *subject* is
+    not modified. This is a much lighter weight approach for limited-use
+    setups than cloning an environment, for example to handle a builder
+    call with keyword arguments that make a temporary change to the
+    current environment::
+
+        env.Program(target="foo", source=sources, DEBUG=True)
+
+    While the majority of methods are proxied from the underlying environment
+    class, enough plumbing is defined in this class for it to behave
+    like an ordinary Environment without the caller needing to know it is
+    "special" in some way.  We don't call the initializer of the class
+    we're proxying, rather depend on it already being properly set up.
+
+    Deletion is handled specially, if a variable was explicitly deleted,
+    it should no longer appear to be in the env, but we also don't want to
+    modify the subject environment.
+
+    :class:`OverrideEnvironment` can nest arbitratily, *subject*
+    can be an existing instance. Although instances can be
+    instantiated directly, the expected use is to call the
+    :meth:`~SubstitutionEnvironment.Override` method as a factory.
+
+    Note Python does not give us a way to assure the subject environment
+    is not modified. Assigning to a variable creates a new entry in
+    the override, but moditying a variable will first fetch the one
+    from the subject, and if mutable, it will just be modified in place.
+    For example: ``over_env.Append(CPPDEFINES="-O")``, where ``CPPDEFINES``
+    is an existing list or :class:`~SCons.Util.CLVar`, will successfully
+    append to ``CPPDEFINES`` in the subject env.  To avoid such leakage,
+    clients such as Scanners, Emitters and Action functions called by a
+    Builder using override syntax must take care if modifying an env
+    (which is not advised anyway) in case they were passed an
+    ``OverrideEnvironment``.
     """
 
-    def __init__(self, subject, overrides=None) -> None:
+    def __init__(self, subject, overrides: Optional[dict] = None) -> None:
         if SCons.Debug.track_instances: logInstanceCreation(self, 'Environment.OverrideEnvironment')
+        overrides = {} if overrides is None else overrides
+        # set these directly via __dict__ to avoid trapping by __setattr__
         self.__dict__['__subject'] = subject
-        if overrides is None:
-            self.__dict__['overrides'] = {}
-        else:
-            self.__dict__['overrides'] = overrides
+        self.__dict__['overrides'] = overrides
+        self.__dict__['__deleted'] = []
 
     # Methods that make this class act like a proxy.
+
     def __getattr__(self, name):
+        # Proxied environment methods don't know they could be called with
+        # us as 'self' and may access the _data consvar dict directly.
```

Comment:
how about instead of `us as 'self'`, `an OverrideEnvironment as 'self'` ?  Or am I misunderstanding this comment?

Reply:
It means - in a function that happens to be written as a method of `SubstitutionEnvironment` or `Base`, you can't make assumptions about what kind of object is passed as `self`, it could easily be an `OverrideEnvironment`, which isn't even in the class hierarchy. Yes, could spell it out.  I'll also have to eventually revise the CHANGES entry, since some of what it says got undone, I just noticed.

Reply:
Updated both.